### PR TITLE
[SPARK-47781][SQL] Handle negative scale decimals for JDBC data sources

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

SPARK-30252 has disabled the definition of the negative scale for decimals. It has a regression that also disabled reading negative scale decimals from JDBC data sources. **Although there is a legacy config to restore the old behavior**, it seemed neither designed for such a case nor convenient in a data pipeline that extracts negative scale decimals from a database such as Oracle to Parquet files w/o negative scale decimal support.

In addition, Postgres has the negative scale decimals support since v15, which was one of the supporters for disabling negative scale decimals on our side.

In this PR, we change the schema from `decimal(p,s)` to `decimal(p-s,0)` if s<0.

### Does this PR introduce _any_ user-facing change?

Negative scale decimals have many supporters for rounding the internal parts, such as Oracle, Postgres, etc.
- Oracle
> Negative scale is the number of significant digits to the left of the decimal point, to but not including the least significant digit. For negative scale the least significant digit is on the left side of the decimal point, because the actual data is rounded to the specified number of places to the left of the decimal point. For example, a specification of (10,-2) means to round to hundreds.

- Postgres
> Beginning in PostgreSQL 15, it is allowed to declare a numeric column with a negative scale. Then values will be rounded to the left of the decimal point. The precision still represents the maximum number of non-rounded digits. Thus, a column declared as
NUMERIC(2, -3)
will round values to the nearest thousand and can store values between -99000 and 99000, inclusive. It is also allowed to declare a scale larger than the declared precision. Such a column can only hold fractional values, and it requires the number of zero digits just to the right of the decimal point to be at least the declared scale minus the declared precision. For example, a column declared as

### How was this patch tested?

new tests

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MySQLIntegrationSuite.scala [hunk 1]
```diff
@@ -139,7 +139,7 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationSuite {
     assert(row.getInt(3) == 77777)
     assert(row.getInt(4) == 123456789)
     assert(row.getLong(5) == 123456789012345L)
-    val bd = new BigDecimal("123456789012345.12345678901234500000")
+    val bd = new BigDecimal("123456789012345.123456789012345000")
```
Comment (yaooqinn) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MySQLIntegrationSuite.scala:142]: FYI, the original value from MySQL is 123456789012345.123456789012345

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 2]
```diff
@@ -196,9 +196,13 @@ object JdbcUtils extends Logging with SQLConfHelper {
     case java.sql.Types.CHAR => CharType(precision)
     case java.sql.Types.CLOB => StringType
     case java.sql.Types.DATE => DateType
-    case java.sql.Types.DECIMAL if precision != 0 || scale != 0 =>
-      DecimalType.bounded(precision, scale)
-    case java.sql.Types.DECIMAL => DecimalType.SYSTEM_DEFAULT
+    case java.sql.Types.DECIMAL | java.sql.Types.NUMERIC if precision != 0 || scale != 0 =>
+      if (scale < 0) {
+        DecimalPrecision.bounded(precision - scale, 0)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:201]: Actually, I think these two changes are orthogonal. To support negative scale, we turn scale to 0, then it doesn't matter how we truncate the precision, as scale is already 0.

For non-negative scale, we can use a different way to truncate the precision. It's better to do it in a separate PR as it's a behavior change.

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:201]: +1, make sense

**Comment by yaooqinn** [comment]
cc @dongjoon-hyun @cloud-fan, PTAL when you have some time

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1 LGTM

**Comment by yaooqinn** [comment]
+1, LGTM.

Thank you. @cloud-fan @dongjoon-hyun
